usr.bin/wc/wc.1
.\"	$NetBSD: wc.1,v 1.20 2026/02/02 21:20:18 kre Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)wc.1	8.2 (Berkeley) 4/19/94
.\"
.Dd September 1, 2019
.Dt WC 1
.Os
.Sh NAME
.Nm wc
.Nd word, line, and byte count
.Sh SYNOPSIS
.Nm
.Op Fl c Ns | Ns Fl m
.Op Fl Llw
.Op Ar file ...
.Sh DESCRIPTION
The
.Nm
utility displays the number of lines, words, bytes and characters contained in each
input
.Ar file
(or standard input, by default) to the standard output.
A line is defined as a string of characters delimited by a <newline>
character,
and a word is defined as a string of characters delimited by white space
characters.
White space characters are the set of characters for which the
.Xr iswspace 3
function returns true.
If more than one input file is specified, a line of cumulative counts
for all the files is displayed on a separate line after the output for
the last file.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl c
The number of bytes in each input file
is written to the standard output.
.It Fl L
The number of characters in the longest line of each input file
is written to the standard output.
.It Fl l
The number of lines in each input file
is written to the standard output.
.It Fl m
The number of characters in each input file
is written to the standard output.
.It Fl w
The number of words in each input file
is written to the standard output.
.El
.Pp
The
.Fl c
and
.Fl m
options are incompatible, the last specified is used.
.Pp
When an option is specified,
.Nm
only
reports the
information requested by that option.
The default action is equivalent to all the flags
.Fl clw
having been specified.
.Pp
The following operands are available:
.Bl -tag -width Ds
.It Ar file
A pathname of an input file.
.El
.Pp
If no file names
are specified, the standard input is used and
no file name is displayed.
.Pp
By default, the standard output contains a line for each
input file of the form:
.Bd -literal -offset indent
lines	 words	bytes	file_name
.Ed
.Sh EXIT STATUS
.Ex -std wc
.Sh COMPATIBILITY
Historically, the
.Nm
utility was documented to define a word as a ``maximal string of
characters delimited by
.Aq space ,
.Aq tab
or
.Aq newline
characters''.
The implementation, however, didn't handle non-printing characters
correctly so that ``  ^D^E  '' counted as 6 spaces, while ``foo^D^Ebar''
counted as 8 characters.
.Bx 4
systems after
.Bx 4.3
modified the implementation to be consistent
with the documentation.
This implementation defines a ``word'' in terms of the
.Xr iswspace 3
function, as required by
.St -p1003.2 .
.Pp
The
.Fl L
option is a non-standard extension, compatible with the
.Fl L
option of the GNU and
.Fx
.Nm
utilities.
.Sh SEE ALSO
.Xr iswspace 3
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.2-92 .
.Sh HISTORY
A
.Nm
utility appeared in
.At v1 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3228 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	man(1): Add -l flag to parse argument as file name.
		[christos 20250902]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	ctype(3): New environment variable LIBC_ALLOWCTYPEABUSE.
		[riastradh 20250915]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]
	gpt(8): Add various riscv-specific partition types. [thorpej 20251011]
	fsck_lfs(8): Enable checking LFS64 volumes. [perseant 20251012]
	m68k: New port: mac68ksf - Macintosh 68k softfloat. [nat 20251020]
	sc16is7xx(4): NXP SC16IX7xx family of UART chips that are behind
		a I2C or SPI bus.  [brad 20251024]
	compat_linux(8): Add support for semtimedop(2), from Ricardo Branco.
		[christos 20251026]
	ftp(1): Add "connect" as a synonym for "open". [lukem 20251028]
	lfs: Add in-kernel cleaner. [perseant 20251106]
	m68k: Experimental new pmap module for systems with 68k MMUs
		(plus the HP MMU). [thorpej 20251108]
	compat_linux(8): Add prctl(2) lwp_{g,s}etname, from Kevin Bloom.
		[christos 20251110]
	sparc64: Add implementation of Sun's VIS instruction API.
		[nia 20251113]
	iconv(1): Support -o outfile like gnu does and check writes for
		failure. [christos 20251121]
	cgd(4): Improve throughput of AES disk encryption using SSE2.
		[riastradh 20251123]
	powerpc: Switch to GCC 14. [martin 20251128]
	dreamcast, atu(4), icp(4), gpsa(8): Relicense various NetBSD
		components from 4-clause to 3-clause BSD. [nia 20251128]
	umcpmio(4): Reorg and support for the MCP2210, USB to SPI
		bridge. [brad 20251129]
	ucom(4): Allow using as boot console. [manu 20251129]
	sparc64: Use 64-bit optimized AES code for disk encryption.
		[nia 20251201]
	iwm(4): Update driver code, improving performance and stability.
		[mlelstv 20251201]
	lfs: Various stability improvements. [perseant 20251202]
	evbppc: Enable PaX security features in WII kernel configuration.
		[nia 20251204]
	viaide(4): Identify and attach VT6420 IDE (PATA) controller.
		[andvar 20251204]
	diff(1), diff3(1): Import BSD-licensed diff to replace GNU
		diffutils, build with MKBSDDIFF=yes. [nia 20251204]
	amd64: Support ramdisks when booting in GENPVH mode.
		[khorben 20251205]
	evbppc: wii: Add support for GameCube controller sockets
		as uhid(4) (now "userspace hid") devices.
		[jmcneill 20251208]
	openresolv(8): import 3.17.1 (patched to 3.17.3) [roy 20251213]
	editline(3): Add a method to set custom word characters,
		from bsiegert. [christos 20251214]
	evbarm: Add support for TI OMAP4 PandaBoard-ES, from Rui-Xiang Guo.
		[skrll 20251216]
	tzcode: Updated to 2025c. [christos 20251218]
	tzdata: Updated to 2025c (using 2025cgtz) [kre 20251219]
	re(4): Made the code MPSAFE with a private mutex.
		[mlelstv 20251220]
	lm(4): Support five aux fan sensors for NCT6794D. [isaki 20251224]
	inetd(8): Add an optional accept limit to stream/nowait services.
		[mlelstv 20251227]
	sun2: Add various MK build knobs and disable them on sun2 to trim
		down the installation. [thorpej 20251228]
	diff(1): Switch the default to MKBSDDIFF=yes. [nia 20260102]
	makefs(8): Honour -m option to set a maximum FS size for
		cd9660 images. [nia 20260107]
	i386, amd64, sparc64: Exclude various base features from the

usr.bin/wc/wc.c
/*	$NetBSD: wc.c,v 1.37 2024/01/14 17:39:19 christos Exp $	*/

/*
 * Copyright (c) 1980, 1987, 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1987, 1991, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)wc.c	8.2 (Berkeley) 5/2/95";
#else
__RCSID("$NetBSD: wc.c,v 1.37 2024/01/14 17:39:19 christos Exp $");
#endif
#endif /* not lint */

/* wc line, word, char count and optionally longest line. */

#include <sys/param.h>
#include <sys/file.h>
#include <sys/stat.h>

#include <ctype.h>
#include <fcntl.h>
#include <err.h>
#include <errno.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <wchar.h>
#include <wctype.h>

#ifdef NO_QUAD
typedef u_long wc_count_t;
# define WCFMT	" %7lu"
# define WCCAST unsigned long
#else
typedef u_quad_t wc_count_t;
# define WCFMT	" %7llu"
# define WCCAST	unsigned long long
#endif

static wc_count_t	tlinect, twordct, tcharct, tlongest;
static bool		doline, doword, dobyte, dochar, dolongest;
static int 		rval = 0;

static void	cnt(const char *);
static void	print_counts(wc_count_t, wc_count_t, wc_count_t, wc_count_t,
		    const char *);
__dead static void	usage(void);
static size_t	do_mb(wchar_t *, const char *, size_t, mbstate_t *,
		    size_t *, const char *);

int
main(int argc, char *argv[])
{
	int ch;

	setlocale(LC_ALL, "");

	while ((ch = getopt(argc, argv, "lwcmL")) != -1)
		switch (ch) {
		case 'l':
			doline = true;
			break;
		case 'w':
			doword = true;
			break;
		case 'm':
			dochar = true;
			dobyte = 0;
			break;
		case 'c':
			dochar = 0;
			dobyte = true;
			break;
		case 'L':
			dolongest = true;
			break;
		case '?':
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	/* Wc's flags are on by default. */
	if (!(doline || doword || dobyte || dochar || dolongest))
		doline = doword = dobyte = true;

	if (*argv == NULL) {
		cnt(NULL);
	} else {
		bool dototal = (argc > 1);

		do {
			cnt(*argv);
		} while(*++argv);

		if (dototal) {
			print_counts(tlinect, twordct, tcharct, tlongest,
			    "total");
		}
	}

	exit(rval);
}

static size_t
do_mb(wchar_t *wc, const char *p, size_t len, mbstate_t *st,
    size_t *retcnt, const char *file)
{
	size_t r;
	size_t c = 0;

	do {
		r = mbrtowc(wc, p, len, st);
		if (r == (size_t)-1) {
			warnx("%s: invalid byte sequence", file);
			rval = 1;

			/* XXX skip 1 byte */
			len--;
			p++;
			memset(st, 0, sizeof(*st));
			continue;
		} else if (r == (size_t)-2)
			break;
		else if (r == 0)
			r = 1;
		c++;
		if (wc)
			wc++;
		len -= r;
		p += r;
	} while (len > 0);

	*retcnt = c;

	return (r);
}

static void
cnt(const char *file)
{
	u_char buf[MAXBSIZE];
	wchar_t wbuf[MAXBSIZE];
	struct stat sb;
	wc_count_t charct, linect, wordct, longest;
	mbstate_t st;
	u_char *C;
	wchar_t *WC;
	const char *name;			/* filename or <stdin> */
	size_t r = 0;
	int fd, len = 0;

	linect = wordct = charct = longest = 0;
	if (file != NULL) {
		if ((fd = open(file, O_RDONLY, 0)) < 0) {
			warn("%s", file);
			rval = 1;
			return;
		}
		name = file;
	} else {
		fd = STDIN_FILENO;
		name = "<stdin>";
	}

	if (dochar || doword || dolongest)
		(void)memset(&st, 0, sizeof(st));

	if (!(doword || dolongest)) {
		/*
		 * line counting is split out because it's a lot
		 * faster to get lines than to get words, since
		 * the word count requires some logic.
		 */
		if (doline || dochar) {
			while ((len = read(fd, buf, MAXBSIZE)) > 0) {
				if (dochar) {
					size_t wlen;

					r = do_mb(0, (char *)buf, (size_t)len,
					    &st, &wlen, name);
					charct += wlen;
				} else if (dobyte)
					charct += len;
				if (doline) {
					for (C = buf; len--; ++C) {
						if (*C == '\n')
							++linect;
					}
				}
			}
		}

		/*
		 * if all we need is the number of characters and
		 * it's a directory or a regular or linked file, just
		 * stat the puppy.  We avoid testing for it not being
		 * a special device in case someone adds a new type
		 * of inode.
		 */
		else if (dobyte) {
			if (fstat(fd, &sb)) {
				warn("%s", name);
				rval = 1;
			} else {
				if (sb.st_size != 0 &&
				    (S_ISREG(sb.st_mode) ||
				    S_ISLNK(sb.st_mode) ||
				    S_ISDIR(sb.st_mode))) {
					charct = sb.st_size;
				} else {
					while ((len =
					    read(fd, buf, MAXBSIZE)) > 0)
						charct += len;
				}
			}
		}
	} else {
		/* do it the hard way... */
		wc_count_t linelen;
                bool       gotsp;

		linelen = 0;
		gotsp = true;
		while ((len = read(fd, buf, MAXBSIZE)) > 0) {
			size_t wlen;

			r = do_mb(wbuf, (char *)buf, (size_t)len, &st, &wlen,
			    name);
			if (dochar) {
				charct += wlen;
			} else if (dobyte) {
				charct += len;
			}
			for (WC = wbuf; wlen--; ++WC) {
				if (iswspace(*WC)) {
					gotsp = true;
					if (*WC == L'\n') {
						++linect;
						if (linelen > longest)
							longest = linelen;
						linelen = 0;
					} else {
						linelen++;
					}
				} else {
					/*
					 * This line implements the POSIX
					 * spec, i.e. a word is a "maximal
					 * string of characters delimited by
					 * whitespace."  Notice nothing was
					 * said about a character being
					 * printing or non-printing.
					 */
					if (gotsp) {
						gotsp = false;
						++wordct;
					}

					linelen++;
				}
			}
		}
	}

	if (len == -1) {
		warn("%s", name);
		rval = 1;
	}
	if (dochar && r == (size_t)-2) {
		warnx("%s: incomplete multibyte character", name);
		rval = 1;
	}

	print_counts(linect, wordct, charct, longest, file);

	/*
	 * don't bother checkint doline, doword, or dobyte --- speeds
	 * up the common case
	 */
	tlinect += linect;
	twordct += wordct;
	tcharct += charct;
	if (dolongest && longest > tlongest)
		tlongest = longest;

	if (close(fd)) {
		warn("%s", name);
		rval = 1;
	}
}

static void
print_counts(wc_count_t lines, wc_count_t words, wc_count_t chars,
    wc_count_t longest, const char *name)
{

	if (doline)
		(void)printf(WCFMT, (WCCAST)lines);
	if (doword)
		(void)printf(WCFMT, (WCCAST)words);
	if (dobyte || dochar)
		(void)printf(WCFMT, (WCCAST)chars);
	if (dolongest)
		(void)printf(WCFMT, (WCCAST)longest);

	if (name != NULL)
		(void)printf(" %s\n", name);
	else
		(void)putchar('\n');
}

static void
usage(void)
{

	(void)fprintf(stderr, "usage: wc [-c | -m] [-Llw] [file ...]\n");
	exit(1);
}
